Add Secrets and Vaults documentation (#1505)
ngrok/ngrok-docs: docs/traffic-policy/secrets.mdx

---
title: Traffic Policy Secrets
sidebar_label: Secrets
description: Store sensitive data in encrypted vaults and reference them dynamically in your traffic policies.
---

Secrets enable you to store sensitive data in encrypted vaults and reference them dynamically in your traffic policies.
This feature eliminates the need to hardcode passwords, API keys, and other sensitive values directly in policy YAML files.

:::info

> Secrets and vaults are currently in Developer Preview.

Reviewer: We may be able to document this as 'open beta' (thus eliminating the 'request to join early access' requirement).

I've asked for confirmation we can eliminate this requirement here:
https://ngrok.slack.com/archives/C01CYSSHVFS/p1754513489994749

The ngrok API never returns secret values in response payloads.

Reference secrets in your Traffic Policy using the `secrets.get()` macro:

```yaml
secrets.get("vault-name", "secret-name")
```

The macro dynamically retrieves the secret value from the specified vault at runtime.

## Using vaults and secrets

To use secrets, you'll need to create a vault to store them in.

### Create a vault

Use the ngrok Agent CLI to create a vault:

```shell
ngrok api vaults create --name "my-vault" --description "Description of vault purpose"
```

The response includes the vault ID, which you'll need when creating secrets within the vault:

```json
{
  "id": "vault_2yNPzuk6GjHrx3mlOCkJK42RsdR",
  "name": "my-vault",
  "description": "Description of vault purpose",
  "created_at": "2025-08-01T21:10:37Z",
  "uri": "https://api.ngrok.com/vaults/vault_2yNPzuk6GjHrx3mlOCkJK42RsdR"
}
```

### Create a secret

Create a secret within a vault using the vault ID:

```shell
ngrok api secrets create \
  --name "my-secret" \
  --value "secret-value" \
  --vault-id "vault_2yNPzuk6GjHrx3mlOCkJK42RsdR" \
  --description "Optional description"
```

### Supported actions and fields

> Secrets are supported in the following Traffic Policy actions and fields:

Reviewer: ```suggestion
Secrets are currently supported in the following Traffic Policy actions and fields:
```

I think we might want to allude to the fact that more fields will come later, even if not directly. We aspire to support "CEL everywhere' even if this is not yet the case

> - **basic-auth**: `credentials`

Reviewer: Should we hyperlink over to each of these actions' documentation from here, you think?

- **jwt-validation**:
  - `issuer.allow_list[*].value`
  - `http.tokens[*].name`
  - `http.tokens[*].prefix`
  - `jws.keys[*].sources[*].additional_jkus`
- **oauth**: `client_secret`
- **oidc**: `client_secret`
- **restrict-ips**: `allow`, `deny`
- **set-vars**: `vars`
- **terminate-tls**:
  - `server_private_key`
  - `server_certificate`
  - `mutual_tls_certificate_authorities`
- **verify-webhook**: `secret`

## Common use cases for secrets

### Basic authentication

```yaml
on_http_request:
  - actions:
      - type: basic-auth
        config:
          credentials:

>             - "${secrets.get('my-vault', 'password')}"

Reviewer: ```suggestion
            - "user123:${secrets.get('my-vault', 'password')}"
```

These fields are tuples

Author: Would this also be the usage for the webhook example below? (Is it always a tuple?)

Reviewer: Basic auth is a tuple of username:password. Webhook is just a secret (not a tuple)

```

### Webhook verification

```yaml
on_http_request:
  - actions:
      - type: verify-webhook
        config:
          provider: github
          secret: "${secrets.get('webhooks-vault', 'github-secret')}"
```

## Audit events

### Vault events

| Event              | Description                                                     |
| ------------------ | --------------------------------------------------------------- |
| `vault_created.v0` | Triggered when a vault is created                               |
| `vault_deleted.v0` | Triggered when a vault is deleted                               |
| `vault_updated.v0` | Triggered when a vault is updated (name, description, metadata) |

### Secret events

| Event               | Description                        |
| ------------------- | ---------------------------------- |
| `secret_created.v0` | Triggered when a secret is created |
| `secret_deleted.v0` | Triggered when a secret is deleted |
| `secret_updated.v0` | Triggered when a secret is updated |

All audit events include the full resource details in the logs.

## Limits and pricing

> TK

Reviewer: TK?

| Plan     | Vault Limit                  | Secret Limit               | Price/Vault | Price/Secret |
| -------- | ---------------------------- | -------------------------- | ----------- | ------------ |
| Free     | 5 (hard limit)               | 5 total (hard limit)       | $0.00       | $0.00        |
| Personal | 5 (hard limit)               | 25 total (hard limit)      | $0.00       | $0.00        |
| PayGo    | 5 (soft limit, configurable) | 500 default (configurable) | $0.00       | $0.00        |

> If you need higher limits than your plan provides, contact support.

Reviewer: ```suggestion
If you need higher limits than your plan provides, [contact support](mailto:support@ngrok.com).
```

```

### Webhook verification

```yaml
on_http_request:
  - actions:
      - type: verify-webhook
        config:
          provider: github
          secret: "${secrets.get('webhooks-vault', 'github-secret')}"
```

## Audit events

Secrets and vaults emit the following [audit events](/obs/events/reference/#audit-events), all of which include the full resource details in the logs:

### Vault events

| Event              | Description                                                     |
| ------------------ | --------------------------------------------------------------- |
| `vault_created.v0` | Triggered when a vault is created                               |
| `vault_deleted.v0` | Triggered when a vault is deleted                               |
| `vault_updated.v0` | Triggered when a vault is updated (name, description, metadata) |

### Secret events

| Event               | Description                        |
| ------------------- | ---------------------------------- |
| `secret_created.v0` | Triggered when a secret is created |
| `secret_deleted.v0` | Triggered when a secret is deleted |
| `secret_updated.v0` | Triggered when a secret is updated |

## Limits and pricing

> Secrets and vaults are free to use during Developer Preview and Open Beta.

Author: ```suggestion
:::warning
Secrets and vaults are free to use during the Developer Preview phase.
Please note that this feature _will_ be billed and metered in the future.
:::
```

        config:
          provider: github
          secret: "${secrets.get('webhooks-vault', 'github-secret')}"
```

## Audit events

Secrets and vaults emit the following [audit events](/obs/events/reference/#audit-events), all of which include the full resource details in the logs:

### Vault events

| Event              | Description                                                     |
| ------------------ | --------------------------------------------------------------- |
| `vault_created.v0` | Triggered when a vault is created                               |
| `vault_deleted.v0` | Triggered when a vault is deleted                               |
| `vault_updated.v0` | Triggered when a vault is updated (name, description, metadata) |

### Secret events

| Event               | Description                        |
| ------------------- | ---------------------------------- |
| `secret_created.v0` | Triggered when a secret is created |
| `secret_deleted.v0` | Triggered when a secret is deleted |
| `secret_updated.v0` | Triggered when a secret is updated |

## Limits and pricing

:::warning
Secrets and vaults are free to use during the Developer Preview phase.

> Please note that this feature _will_ be billed and metered in the future.

Reviewer: ```suggestion
Please note that this feature _will_ be billed and metered in the future, per the plan details below.
```

:::

| Plan     | Vault Limit                  | Secret Limit               |
| -------- | ---------------------------- | -------------------------- |
| Free     | 5 (hard limit)               | 5 total (hard limit)       |
| Personal | 5 (hard limit)               | 25 total (hard limit)      |
| PayGo    | 5 (soft limit, configurable) | 500 default (configurable) |

> If you need higher limits than your plan provides, [contact support](mailto:support@ngrok.com).

Author: ```suggestion
Please [contact support](mailto:support@ngrok.com) to request higher limits than your plan provides, or to configure limits for the PayGo plan.
```
